src/Sql/SqlMysql.php

```diff
@@ -107,7 +108,21 @@ public function createdbSql($dbname, $quoted = false)
             $user = sprintf("'%s'@'%s'", $dbSpec['username'], $domain);
             $sql[] = sprintf("DROP USER IF EXISTS %s;", $user);
             $sql[] = sprintf("CREATE USER %s IDENTIFIED WITH mysql_native_password;", $user);
-            $sql[] = sprintf("SET PASSWORD FOR %s = PASSWORD('%s');", $user, $dbSpec['password']);
+
+            // For MariaDB, ALTER USER was introduced in 10.2 so before that use SET PASSWORD.
+            // @see Drupal\Core\Database\Driver\mysql\Connection::getMariaDbVersionMatch()
+            // in Drupal 9.
+            $this->su();
```

Comment:
This is already handled by the caller. See \Drush\Sql\SqlBase::createdb
```suggestion
```

Reply:
Unfortunately that doesn't work - createdb does it too late.  We need to have su now for the query `SELECT VERSION();` that is run immediately.

Reply:
Is it possible to determine version without admin perms? That would be generally useful. We could then make a sql SqlMariaDb class that inherits from SqlMysql and specializes createDb accordingly. Very few sites have entered `su` creds because they dont need to create a DB.

Reply:
In general yes we can determine version without admin perms and it would be much easier than the code here because we could call `Drupal\Core\Database\Driver\mysql\Connection::getMariaDbVersionMatch()` (as soon as D8 is no longer supported.)  The only reason we need su now is because the database and user haven't been created yet - we are about to create them.  This code is inside `if (isset($db_superuser))` so it won't trouble all the sites that have not entered `su` creds.

Reply:
This whole complex branch of code is only needed until EOL of D8 - after that the minimum maria version of D9 is already above the problem point.  Seeing as hardly anyone runs this code (like you say they mostly don't enter DB creds) and it took months for anyone to notice it was broken, we could just forget about it.  We could stick with ALTER USER and it's hard luck for anyone on the very old maria version.

Reply:
ALTER USER is needed by recent versions of mariadb and we dont use ALTER USER anywhere, right? So we cant just ignore this issue, if I'm reading this right.

Reply:
Correct we cannot ignore this issue.  The correct patch that works for all database versions supported in Drupal 9 is:

```
-            $sql[] = sprintf("SET PASSWORD FOR %s = PASSWORD('%s');", $user, $dbSpec['password']);
+           $sql[] = sprintf("ALTER USER %s IDENTIFIED BY '%s';", $user, $dbSpec['password']);
```

Everything else is only there for old Maria versions still supported in D8.  If we add it now we would presumably then remove it again in 10 months time.  Please let me know if you would prefer to commit the simple patch now.